ethstats/ethstats.go

```diff
@@ -211,93 +253,215 @@ func (s *Service) loop() {
 				case txCh <- struct{}{}:
 				default:
 				}
-
 			// node stopped
 			case <-txSub.Err():
 				break HandleLoop
 			case <-headSub.Err():
 				break HandleLoop
+			case delegateSignMsg := <-istDelegateSignCh:
+				var statsPayload StatsPayload
+				err := json.Unmarshal(delegateSignMsg.Payload, &statsPayload)
+				if err == nil {
+					if s.backend.ProxyNode() != nil {
+						select {
+						case signCh <- statsPayload:
+						default:
+						}
+					}
+					if s.backend.ProxiedPeer() != nil {
+						select {
+						case sendCh <- statsPayload:
+						default:
+						}
+					}
+				}
 			}
 		}
 		close(quitCh)
 	}()
+
 	// Loop reporting until termination
 	for {
-		// Resolve the URL, defaulting to TLS, but falling back to none too
-		path := fmt.Sprintf("%s/api", s.host)
-		urls := []string{path}
+		if s.backend.ProxyNode() != nil {
+			messageToSign := <-signCh
+			if err := s.handleDelegateSign(messageToSign); err != nil {
+				log.Warn("Delegate sign failed", "err", err)
+			}
+		} else {
+			conn, err := s.getConnection()
+			if err != nil {
+				log.Warn("Stats server unreachable", "err", err)
+				time.Sleep(10 * time.Second)
+				continue
+			}
+			// Authenticate the client with the server
+			if err = s.login(conn, sendCh); err != nil {
+				log.Warn("Stats login failed", "err", err)
+				conn.Close()
+				time.Sleep(10 * time.Second)
+				continue
+			}
+			go s.readLoop(conn)
 
-		if !strings.Contains(path, "://") { // url.Parse and url.IsAbs is unsuitable (https://github.com/golang/go/issues/19779)
-			urls = []string{"wss://" + path, "ws://" + path}
-		}
-		// Establish a websocket connection to the server on any supported URL
-		var (
-			conf *websocket.Config
-			conn *websocket.Conn
-			err  error
-		)
-		for _, url := range urls {
-			if conf, err = websocket.NewConfig(url, "http://localhost/"); err != nil {
+			// Send the initial stats so our node looks decent from the get go
+			if err = s.report(conn, sendCh); err != nil {
+				log.Warn("Initial stats report failed", "err", err)
+				conn.Close()
 				continue
 			}
-			conf.Dialer = &net.Dialer{Timeout: 5 * time.Second}
-			if conn, err = websocket.DialConfig(conf); err == nil {
-				break
+			// Keep sending status updates until the connection breaks
+			fullReport := time.NewTicker(15 * time.Second)
```

Comment:
lots of constants, like here & the time.Sleep calls, consider putting them in a const at the top of the file